src/backend/distributed/metadata/pg_get_object_address_12_13_14.c

```diff
@@ -489,9 +537,251 @@ ErrorIfCurrentUserCanNotDistributeObject(ObjectType type, ObjectAddress *addr,
 	if (!skipAclCheck && aclMaskResult == ACL_NO_RIGHTS)
 	{
 		ereport(ERROR, (errmsg("Current user does not have required privileges "
-							   "on %d with type id %d to distribute it", idToCheck,
-							   type)));
+							   "on %d with type id %d to distribute it",
+							   idToCheck, type)));
+	}
+}
+
+
+/*
+ * Copied from PG code.
+ *
+ * Policy function: is the given extension trusted for installation by a
+ * non-superuser?
+ *
+ * (Update the errhint logic below if you change this.)
+ */
+static bool
+extension_is_trusted(ExtensionControlFile *control)
+{
+	/* Never trust unless extension's control file says it's okay */
+	if (!control->trusted)
+	{
+		return false;
+	}
+
+	/* Allow if user has CREATE privilege on current database */
+	AclResult aclresult = pg_database_aclcheck(MyDatabaseId, GetUserId(), ACL_CREATE);
+	if (aclresult == ACLCHECK_OK)
+	{
+		return true;
+	}
+	return false;
+}
+
+
+/*
+ * Copied from PG code.
+ *
+ * Read the primary control file for the specified extension.
+ */
+static ExtensionControlFile *
+read_extension_control_file(const char *extname)
+{
+	/*
+	 * Set up default values.  Pointer fields are initially null.
+	 */
+	ExtensionControlFile *control =
+		(ExtensionControlFile *) palloc0(sizeof(ExtensionControlFile));
+	control->name = pstrdup(extname);
+	control->relocatable = false;
+	control->superuser = true;
+	control->trusted = false;
+	control->encoding = -1;
+
+	/*
+	 * Parse the primary control file.
+	 */
+	parse_extension_control_file(control, NULL);
+
+	return control;
+}
+
+
+/*
+ * Copied from PG code.
+ *
+ * Parse contents of primary or auxiliary control file, and fill in
+ * fields of *control.  We parse primary file if version == NULL,
+ * else the optional auxiliary file for that version.
+ *
+ * Control files are supposed to be very short, half a dozen lines,
+ * so we don't worry about memory allocation risks here.  Also we don't
+ * worry about what encoding it's in; all values are expected to be ASCII.
+ */
+static void
+parse_extension_control_file(ExtensionControlFile *control,
```

Comment:
Instead of copy & pasting these codes, we might consider relying on `pg_available_extension_versions `. For example, in `AvailableExtensionVersion` we rely on `pg_available_extensions`, we could have a similar function? 

I guess it is nicer to rely on PG instead of copy & paste code?